lib/api/model/model.dart

```diff
@@ -698,3 +710,82 @@ class DmMessage extends Message {
   @override
   Map<String, dynamic> toJson() => _$DmMessageToJson(this);
 }
+
+enum MessageEditState {
+  none,
+  edited,
+  moved;
+
+  // Adapted from the shared code:
+  //   https://github.com/zulip/zulip/blob/1fac99733/web/shared/src/resolved_topic.ts
+  // The canonical resolved-topic prefix.
+  static const String _resolvedTopicPrefix = '✔ ';
+
+  static bool _isResolved(String topic) {
+    return topic.startsWith(_resolvedTopicPrefix);
+  }
+
+  /// Whether two topics are equal, ignoring any resolved-topic prefix.
+  ///
+  /// When a topic is resolved, the clients agree on adding a ✔ prefix to the
+  /// topic string. Topics whose only difference is the ✔ prefix are considered
+  /// the same. This helper can be helpful when checking if a message has been
+  /// moved.
+  ///
+  /// TODO(#744) Extract to its own home when supporting "mark as resolve".
+  static bool sameTopicModuloResolve(String topic, String prevTopic) {
+    if (_isResolved(topic) && topic.substring(_resolvedTopicPrefix.length) == prevTopic) {
```

Comment:
nit: As is, this logic only really makes sense if you go look at the implementation of `_isResolved` and see that it's checking the topic starts with `_resolvedTopicPrefix` — the second condition doesn't make much sense to ask until one knows that starts-with condition is true.

IOW, for the reader the code only really makes much sense once they mentally inline the definition of `_isResolved`. So it's clearer if we just do that in the first place:
```suggestion
    if (topic.startsWith(_resolvedTopicPrefix)
      && topic.substring(_resolvedTopicPrefix.length) == prevTopic) {
```